analyzers/src/SonarAnalyzer.CSharp/Extensions/AssignmentExpressionSyntaxExtensions.cs

```diff
@@ -26,8 +26,114 @@
+
+            static bool MapTupleElements(ImmutableArray<AssignmentMapping>.Builder arrayBuilder, TupleExpressionSyntaxWrapper left, TupleExpressionSyntaxWrapper right)
+            {
+                if (left.Arguments.Count != right.Arguments.Count)
+                {
+                    return false;
+                }
+
+                var leftEnumerator = left.Arguments.GetEnumerator();
+                var rightEnumerator = right.Arguments.GetEnumerator();
+                while (leftEnumerator.MoveNext() && rightEnumerator.MoveNext())
+                {
+                    var leftExpression = leftEnumerator.Current.Expression;
+                    var rightExpression = rightEnumerator.Current.Expression;
+                    if (TupleExpressionSyntaxWrapper.IsInstance(leftExpression)
+                        && TupleExpressionSyntaxWrapper.IsInstance(rightExpression))
+                    {
+                        if (!MapTupleElements(arrayBuilder, (TupleExpressionSyntaxWrapper)leftExpression, (TupleExpressionSyntaxWrapper)rightExpression))
+                        {
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        arrayBuilder.Add(new AssignmentMapping(leftExpression, rightExpression));
+                    }
+                }
+
+                return true;
+            }
+
+            static bool MapDesignationElements(ImmutableArray<AssignmentMapping>.Builder arrayBuilder, ParenthesizedVariableDesignationSyntaxWrapper left, TupleExpressionSyntaxWrapper right)
+            {
+                if (left.Variables.Count != right.Arguments.Count)
+                {
+                    return false;
+                }
+
+                var leftEnumerator = left.Variables.GetEnumerator();
+                var rightEnumerator = right.Arguments.GetEnumerator();
+                while (leftEnumerator.MoveNext() && rightEnumerator.MoveNext())
+                {
+                    var leftVar = leftEnumerator.Current;
+                    var rightExpression = rightEnumerator.Current.Expression;
+                    if (ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(leftVar)
+                        && TupleExpressionSyntaxWrapper.IsInstance(rightExpression))
+                    {
+                        if (!MapDesignationElements(arrayBuilder, (ParenthesizedVariableDesignationSyntaxWrapper)leftVar, (TupleExpressionSyntaxWrapper)rightExpression))
+                        {
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        arrayBuilder.Add(new AssignmentMapping(leftVar.SyntaxNode, rightExpression));
+                    }
+                }
+
+                return true;
+            }
+        }
+
```

Comment:
AssignmentTargets,  tuple.AllArguments(), declaration.Designation.AllVariables() are also helpers that came in #5770

Reply:
The documentation for `assignment` and `returns` doesn't bring value => should be removed.

Reply:
I don't see a reados to prefer `CSharpSyntaxNode` over `SyntaxNode`. We can use just `SyntaxNode` everywhere.

Reply:
AllArguments: Documentation is not needed at all there, it's obvious from the name.

Reply:
AllArguments: `CollectTupleElements` shouldn't be static and `builder` shouldn't be shadowed.

Reply:
AllVariables: `param` and `returns` documentation is noise and should be removed. `Summary` is useful.

Reply:
AllVariables: `CollectVariables` should not be static and `builder` should not be shadowed.

Reply:
AllVariables: I wouldn't rename the expression to `parenthesized` and use it directly in the `foreach` instead. The cast there is just a syntax sugar and naming it doesn't bring value

Reply:
>Even the flattening part? I don't think this is obvious.

Yes, because it says `All`. I would expect `something.Arguments()` to return unflattened list. You can consider renaming it to `FlattenedArguments` to be explicit (I do not insist). 

The performance cost for capturing classes is insignificant. We have it for every lambda we use, see [sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAOgDIEsB2BHA3AWACgoBmBAJjgGE4BvIuRhMqAFjgHV0AXACyoCGAB24BXAE4BTABRZuAbQC6cADYBnbgEo6DJnoBuA8XADGwsVLgBeON3GjJBQnr3ruyDr0lTpAD2sAfKbmEpKauC5MuowAvkQxQA==).

Reply:
CSharpSyntaxNode  vs another PR: OK

AllVariables: docs for `param` and `returns` were not deleted